enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,553 @@
+
+#### Resource changes
+
+#### Kubelet
+
+In OpenShift, `machine-config-operator` manages `kubelet` configuration.
+Based on the [Infrastructure value](https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go) MCO currently sets the correct `cloud-config` value.
+To support external providers, MCO should not set the `cloud-config` value anymore.
+
+Additionally, it must set the value of the `cloud-provider` [flag](https://github.com/openshift/machine-config-operator/blob/2a51e49b0835bbd3ac60baa685299e86777b064f/templates/worker/01-worker-kubelet/_base/units/kubelet.service.yaml#L32) to `external`.
+This informs `kubelet` that an out-of-tree provider will finish the `Node` initialization process.
+
+While `kubelet` itself does not directly contact or interact with the `CCM`, formation of a functional `Node` in Kubernetes does rely on the `CCM`.
+With in-tree providers, `kubelet` looks up Node IP Addresses, Instance IDs and other metadata directly from the cloud provider and places this information onto the `Node` object.
+
+When using out-of-tree, the `kubelet` taints the `Node` with `node.cloudprovider.kubernetes.io/uninitialized: true` and does not look up any information.
+It is then the responsibility of the `CCM` to look up the Node IP Addresses, Instance IDs and other metadata and labels, place the information onto the `Node` object and then remove the taint from the `Node`.
+
+The information previously set by `kubelet`, now `CCM`, is required to make a `Node` schedulable.
+Without this information, the `scheduler` cannot schedule `Pods` that have any sort of affinity that relies on `Node` metadata (eg failure domain).
+
+#### Kube Controller Manager and Kube API Server
+
+In OpenShift, `kube-controller-manager` and `kube-apiserver` are managed by `cluster-kube-controller-manager-operator` and `cluster-kube-apiserver-operator` respectively.
+When switching to out-of-tree providers, the `cloud-provider` option [should be removed](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager) from their deployments.
+
+##### Kube Controller Manager changes to prolong support for in-tree storage
+
+In the first phase of the transition all cloud providers will still rely on the in-tree storage. `--external-cloud-volume-plugin=<some-provider>` set on `KCM` pod will preserve the in-tree storage controller loops.
```

Comment:
i find this a little confusing, is this saying that `--external-cloud-volume-plugin=KCM` will preserve the in-tree controller?

Reply:
--external-cloud-volume-plugin=aws or some other cloud provider. It will preserve in-tree persistent volume controllers. I'll make some changes to the text.